src/Pact/Compile.hs

```diff
@@ -278,11 +283,66 @@ moduleForm = do
     TTable {} -> return []
     TUse {} -> return []
     TBless {..} -> return [_tBlessed]
-    _ -> syntaxError "Only defun, defpact, defconst, deftable, use, bless allowed in module"
+    TImplements{} -> return []
+    t -> syntaxError $ "Invalid declaration in module scope: " ++ abbrev t
+  let interfaces = flip foldMap bd $ \d -> case d of
+        TImplements{..} -> [_tInterfaceName]
+        _ -> []
   return $ TModule
-    (Module modName (KeySetName keyset) m code modHash blessed)
+    (Module modName (KeySetName keyset) m code modHash blessed interfaces)
     (abstract (const Nothing) (TList bd TyAny bi)) i
 
+implements :: Compile (Term Name)
+implements = do
+  modName <- currentModule'
+  -- TODO: ModuleName should be renamed to something that reflects interface AND module
+  ifName <- (ModuleName . _atomAtom) <$> bareAtom
+  info <- contextInfo
+  return $ TImplements ifName modName info
+  
+interface :: Compile (Term Name)
+interface = do
+  iname' <- _atomAtom <$> bareAtom
+  m <- meta
+  use (psUser . csModule) >>= \ci -> case ci of
+    Just {} -> syntaxError "invalid nested interface or module"
+    Nothing -> return ()
+  info <- contextInfo
+  let code = case info of
+        Info Nothing -> "<code unavailable>"
+        Info (Just (c,_)) -> c
+      -- TODO: ModuleName should be renamed to something that reflects interface AND module
+      iname = ModuleName iname'
+      ihash = hash $ encodeUtf8 (_unCode code)
+  (psUser . csModule) .= Just (iname, ihash)
+  (defs, defInfo) <- interfaceForm
+  eof
+  return $ TModule
+    (Interface iname code m)
+    (abstract (const Nothing) (TList defs TyAny defInfo)) info
+    
+interfaceForm :: Compile ([Term Name], Info)
+interfaceForm = (,) <$> some interfaceForms <*> contextInfo
+  where
+    interfaceForms = withList' Parens $ do 
+      AtomExp{..} <- bareAtom
+      case _atomAtom of
+        "defun" -> commit >> emptyDef 
+        "defconst" -> commit >> defconst
+        _ -> expected "empty form"
```

Comment:
This error is wrong, use error from module def: `t -> syntaxError $ "Invalid declaration in interface scope: " ++ abbrev t`

Reply:
`t` in this case is `Text`, so I can't use `abbrev` at this level. I can, however, use `unpack t`, and just give the same error a la `t -> syntaxError $ "Invalid declaration in interface scope: " ++ unpack t`.